samples: net: http_server: demonstrate dynamic GET and POST

mrodgers-witekio opened the pull request:
Previously the dynamic resource in the http_server sample simply echo'd the received data back to the sender. This doesn't demonstrate how the dynamic handlers would be used in a more typical use case, where the application may need to wait for the full request before acting on it or creating a response (mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/76280)

I've replaced the existing dynamic resource with two new dynamic resources:
- The '/led' resource can be used to turn an LED on the board on or off via a POST request. 
- The '/uptime' resource can be used to query the kernel uptime of the board.

The static resources are updated with some javascript to support demonstrating the new dynamic resources via a web browser.

The web page presented to the user now looks like this:
![image](https://github.com/user-attachments/assets/89f0abb1-293d-4770-ab30-646b9102d854)

mrodgers-witekio committed:
samples: net: http_server: demonstrate dynamic GET and POST

Previously the dynamic resource in the http_server sample simply echo'd
the received data back to the sender. This doesn't demonstrate how the
dynamic handlers would be used in a more typical use case, where the
application may need to wait for the full request before acting on it or
creating a response.

Therefore two new dynamic resources have bee added: a '/led' resource
which can be used to control the board's LED (if present) via a POST
request, and a '/uptime' resource which can be used to query the kernel
uptime via a GET request.

The static resources are updated with some javascript to support
demonstrating the new dynamic resouces via a web browser.

Signed-off-by: Matt Rodgers <mrodgers@witekio.com>

mrodgers-witekio commented on samples/net/sockets/http_server/src/static_web_resources/main.js:
```diff
@@ -0,0 +1,45 @@
+async function fetchUptime()
```
Should we add license headers in a javascript file? I couldn't find any existing `.js` files in the tree (understandably!)

mrodgers-witekio replied:
Never mind, CI answered the question for me

kartben commented on samples/net/sockets/http_server/src/static_web_resources/index.html:
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<html>
```
proposing a couple changes to make sure it's valid HTML (this one is mostly a warning but the changes to input tags are fixing actual invalid HTML)

```suggestion
<html lang="en">
```

kartben commented on samples/net/sockets/http_server/src/static_web_resources/index.html:
```diff
@@ -0,0 +1,20 @@
+
+    <h4>LED Control</h4>
+    <p>The buttons below toggle the LED on the board, or print a message to the console if the board does not have an LED. This demonstrates dynamic handling of a POST request.</p>
+    <input id="led_on" type="button" value="LED on" \>
+    <input id="led_off" type="button" value="LED off" \>
```
```suggestion
    <input id="led_on" type="button" value="LED on">
    <input id="led_off" type="button" value="LED off">
```

kartben commented on samples/net/sockets/http_server/src/main.c:
```diff
@@ -37,59 +44,139 @@ struct http_resource_detail_static index_html_gz_resource_detail = {
+	}
+}
+
+static void parse_led_post(uint8_t *buf, size_t len)
+{
+	if ((len == sizeof("led=on") - 1) && (0 == memcmp("led=on", buf, len))) {
+		led_control(true);
+	} else if ((len == sizeof("led=off") - 1) && (0 == memcmp("led=off", buf, len))) {
+		led_control(false);
+	} else {
+		LOG_WRN("Unexpected POST payload");
+	}
+}
```
I feel like it would be kind of neat to use this as an excuse to show a JSON payload being parsed, but also OK to keep as-is

related, @rlubos @jukkar the sample enables `CONFIG_JSON_LIBRARY` but it looks like maybe it's a leftover since I don't think it's needed (yet! 😄)

mrodgers-witekio replied:
Perhaps a payload like:
```
{
  led_num: 0,
  led_state: true,
}
```
then we can keep the web interface the same (just always set `led_num` to 0), but if you send a POST via `curl` or similar you can control multiple LEDs (if they exist on the board)?

Shouldn't be much more complicated to implement than what I've done already.

jukkar replied:
> the sample enables `CONFIG_JSON_LIBRARY` but it looks like maybe it's a leftover since I don't think it's needed

Yes, that is a historical artifact that is indeed not needed atm.

mrodgers-witekio replied:
JSON payload added for the POST request now.

mrodgers-witekio commented on samples/net/sockets/http_server/src/main.c:
```diff
@@ -37,59 +44,139 @@ struct http_resource_detail_static index_html_gz_resource_detail = {
+};
 
-static uint8_t recv_buffer[1024];
+static int uptime_handler(struct http_client_ctx *client, enum http_data_status status,
```
I've never used websockets before, but sounds like a good excuse to learn if I can find some time!

Maybe some of the network statistics (`CONFIG_NET_STATISTICS`) could be a good option for some different data to send via the websocket connection?

jukkar replied:
The network statistics sending is a good idea, you probably need to encapsulate the stats data as a json object to allow easier parsing at the other end.

mrodgers-witekio replied:
I have some network statistics via websocket working now, it was actually quite straightforward in the end (has me questioning why I've been using Javascript polling all this time :smile: ). See: https://github.com/mrodgers-witekio/zephyr/tree/http_server_sample_websocket_netstats

@kartben @jukkar @rlubos any preference on whether I include it in this PR or make a separate one?

Web page now looks like this:
![image](https://github.com/user-attachments/assets/c8d4388f-0dd8-4114-8d78-0b542e47a51b)

rlubos replied:
If it's a separate commit, I don't mind including it here.

jukkar replied:
I agree, looks like a nice addition to this PR.

kartben replied:
Careful guys, someone might end up asking for real-time charts of the network statistics next :) 🙈

jukkar replied:
> someone might end up asking for real-time charts of the network statistics next

The answer to that is "patches are welcome" so no worries :sunglasses:

jukkar commented on samples/net/sockets/http_server/src/main.c:
```diff
@@ -37,59 +44,139 @@ struct http_resource_detail_static index_html_gz_resource_detail = {
-static int dyn_handler(struct http_client_ctx *client, enum http_data_status status,
+	switch (status) {
+	case HTTP_SERVER_DATA_ABORTED:
+	{
```
Minor style nit, please place the starting `{` to the previous line here and the lines below

jukkar commented on samples/net/sockets/http_server/src/not_found_page.html:
```diff
@@ -1,10 +0,0 @@
-<!DOCTYPE html>
```
Why the not found page was removed?

mrodgers-witekio replied:
I couldn't see the generated `not_found_page.html.gz.inc` file referenced anywhere in the sample.

I've now gone back and had another look, and it seems the file is referenced in `zephyr/subsys/net/lib/http/http_server_http2.c`, but only if `INCLUDE_HTML_CONTENT` is defined. This currently isn't defined anywhere (there's a CMake option `INCLUDE_HTML_CONTENT` but this doesn't set the preprocessor definition)
```
static const char content_404[] = {
#ifdef INCLUDE_HTML_CONTENT
#include "not_found_page.html.gz.inc"
#endif
};
```

jukkar replied:
Ok, some piece of the puzzle is now missing, I think this used to work earlier. Anyway, either we remove the "not found" file + code in http2, or this needs to be fixed properly, but both cases should be done outside of this pr.

mrodgers-witekio replied:
Maybe a Kconfig option in the HTTP server lib to specify a file path to the user's custom 404 page might be clearer? A bit like how `CONFIG_MCUBOOT_SIGNATURE_KEY_FILE` works.

rlubos replied:
One of the awkwardness we've left out from the initial server code, yes I think it'd be better to have this configurable with Kconfig.

rlubos commented on samples/net/sockets/http_server/src/main.c:
```diff
@@ -37,59 +44,139 @@ struct http_resource_detail_static index_html_gz_resource_detail = {
 	.user_data = NULL,
 };
 
-HTTP_RESOURCE_DEFINE(dyn_resource, test_http_service, "/dynamic",
```
Can we keep the old "dynamic" resource in the sample? Maybe rename it to "echo" or something to avoid confusion. It may indeed not be the best demo, but it was actually pretty useful for testing large transfers, would be good to have something like this at hand.

mrodgers-witekio replied:
Added back in as `echo_resource`

rlubos reviewed (commented):
Pretty nice demo!

mrodgers-witekio committed:
samples: net: http_server: use JSON payload for dynamic POST

Accept JSON payload including LED number and state, to demonstrate JSON
parsing and allow control of multiple LEDs if the board supports this.

Signed-off-by: Matt Rodgers <mrodgers@witekio.com>

mrodgers-witekio committed:
samples: net: http_server: sent net stats to client via websocket

Added a websocket connection sending network statistics to a client, to
demonstrate how this can be used to push data to the client as an
alternative to GET requests triggered by a Javascript timer.

Signed-off-by: Matt Rodgers <mrodgers@witekio.com>

kartben reviewed (approved):
This is really cool, thanks a lot for the great work @mrodgers-witekio!
I just tested this successfully on an esp32s3 :)
